### PR TITLE
Detect lost connection and reconnect 

### PR DESCRIPTION
**Pull request by Lemocin**
Issue described in #7 
As i'm not very experienced with asyncio, i don't know if my way is the best solution, but it works for me and a friend of mine.

### CODE REVIEW

File: sml/asyncio.py [hunk 1]
```diff
@@ -34,6 +35,7 @@
 
 class SmlProtocol(SmlBase, asyncio.Protocol):
     _BAUD_RATE = 9600
+    _LAST_UPDATE = 0
```
Comment (mtdcr) [inline sml/asyncio.py]: This should be an instance variable (set `self._last_update = 0` in `def __init__(self)`)

File: sml/asyncio.py [hunk 2]
```diff
@@ -56,6 +59,7 @@ def _delay_reading(self, delay):
     def data_received(self, data: bytes):
         self._buf += data
         delay = 0.5
+        self._LAST_UPDATE = time.time()
```
Comment (mtdcr) [inline sml/asyncio.py]: self._last_update

File: sml/asyncio.py [hunk 3]
```diff
@@ -126,10 +131,23 @@ async def connect(self, loop=None):
...
     async def _disconnect(self):
+        if self._watchdog and not self._watchdog.done():
+            self._watchdog.cancel()
         if self._transport:
             self._transport.abort()
             self._transport = None
 
+    async def _timeout(self):
+        timeout_delay = 120
+        self._LAST_UPDATE = time.time() + timeout_delay
```
Comment (mtdcr) [inline sml/asyncio.py]: This line should not be here. If you need to initialize `_last_update` to something different than 0, you can do it inside `__init__`, for example. `_timeout()` should only read this value.

File: sml/asyncio.py [hunk 4]
```diff
@@ -126,10 +131,23 @@ async def connect(self, loop=None):
...
 
     async def _disconnect(self):
+        if self._watchdog and not self._watchdog.done():
+            self._watchdog.cancel()
         if self._transport:
             self._transport.abort()
             self._transport = None
 
+    async def _timeout(self):
+        timeout_delay = 120
```
Comment (mtdcr) [inline sml/asyncio.py]: This value should probably be configurable.

File: sml/asyncio.py [hunk 5]
```diff
@@ -126,10 +131,23 @@ async def connect(self, loop=None):
...
+        self._LAST_UPDATE = time.time() + timeout_delay
+        while True:
+            await asyncio.sleep(30)
```
Comment (mtdcr) [inline sml/asyncio.py]: Why 30s and not until self._last_update + timeout_delay?

**Review by mtdcr — CHANGES_REQUESTED** [review]
I'm not totally against it, until there's a better solution (see my comment in #7), but there are minor issues to fix before it can get merged.

File: sml/asyncio.py (hunk 4 above)
Comment (Lemocin) [reply sml/asyncio.py]: I have introduced a preconfigured variable within `__init__` which contains a time to wait. If this is 0 then the watchdog will not be started.

File: sml/asyncio.py (hunk 3 above)
Comment (Lemocin) [reply sml/asyncio.py]: Thank you. This is a very good point

File: sml/asyncio.py (hunk 5 above)
Comment (Lemocin) [reply sml/asyncio.py]: I changed it to the timeout_delay. 30 seconds were my initial delay while testing. `self._last_update + timeout_delay` would be too long i think

**Comment by Lemocin** [comment]
> I'm not totally against it, until there's a better solution (see my comment in #7), but there are minor issues to fix before it can get merged.

Thank you very much for the feedback. I tried to resolve them.

File: sml/asyncio.py [hunk 6]
```diff
@@ -126,10 +133,21 @@ async def connect(self, loop=None):
...
     async def _disconnect(self):
+        if self._watchdog and not self._watchdog.done():
+            self._watchdog.cancel()
         if self._transport:
             self._transport.abort()
             self._transport = None
 
+    async def _timeout(self):
+        while True:
+            await asyncio.sleep(self._timeout_delay)
```
Comment (mtdcr) [inline sml/asyncio.py]: I was thinking about something like this:
```python3
sleep_time = self._last_update + self._timeout_delay - time.time()
if sleep_time > 0:
    await asyncio.sleep(sleep_time)
```
or 
```python3
last_update = self._last_update
sleep_time = last_update + self._timeout_delay - time.time()
if sleep_time > 0:
    await asyncio.sleep(sleep_time)
    if last_update != self._last_update:
        continue
logger.warning("Timeout ...")
```

Comment (Lemocin) [reply sml/asyncio.py]: I really like the second version. But with this i need to re-set last_update everytime the connection is made. The reason why it was hardcoded initially is, that after a reconnect it should not get reconnected immediately. 
With the current commit it should work as expected and also uses your second suggestion. Thank you!

**Comment by mtdcr** [comment]
Looks good now, thank you @Lemocin!

**Pull request merged**